controller/snapshot_controller.go

```diff
@@ -0,0 +1,543 @@
+	if !ok {
+		if !requestCreateNewSnapshot || alreadyCreatedBefore {
+			// The snapshotInfo exists inside engine.Status.Snapshots before but disappears now.
+			// Mark snapshotCR as lost track of the corresponding snapshotInfo
+			log.Infof("snapshot CR %v lost track of its snapshotInfo", snapshot.Name)
+			snapshot.Status.Error = "lost track of the corresponding snapshot info inside volume engine"
+		}
+		// Newly created snapshotCR, wait for the snapshotInfo to be appeared inside engine.Status.Snapshot
+		snapshot.Status.ReadyToUse = false
+		return nil
+	}
+
+	if err := syncSnapshotWithSnapshotInfo(snapshot, snapshotInfo, engine.Spec.VolumeSize); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (sc *SnapshotController) generatingEventsForSnapshot(existingSnapshot, snapshot *longhorn.Snapshot) {
+	if !existingSnapshot.Status.MarkRemoved && snapshot.Status.MarkRemoved {
+		sc.eventRecorder.Event(snapshot, v1.EventTypeWarning, "SnapshotDelete", "snapshot is marked as removed")
+	}
+	if snapshot.Spec.CreateSnapshot && existingSnapshot.Status.CreationTime == nil && snapshot.Status.CreationTime != nil {
+		sc.eventRecorder.Eventf(snapshot, v1.EventTypeNormal, "SnapshotCreate", "successfully provisioned the snapshot")
+	}
+	if snapshot.Status.Error != "" && existingSnapshot.Status.Error != snapshot.Status.Error {
+		sc.eventRecorder.Eventf(snapshot, v1.EventTypeWarning, "SnapshotError", "%v", snapshot.Status.Error)
+	}
+	if existingSnapshot.Status.ReadyToUse != snapshot.Status.ReadyToUse {
+		if snapshot.Status.ReadyToUse {
+			sc.eventRecorder.Eventf(snapshot, v1.EventTypeNormal, "SnapshotUpdate", "snapshot becomes ready to use")
+		} else {
+			sc.eventRecorder.Eventf(snapshot, v1.EventTypeWarning, "SnapshotUpdate", "snapshot becomes not ready to use")
+		}
+	}
+}
+
+func syncSnapshotWithSnapshotInfo(snap *longhorn.Snapshot, snapInfo *longhorn.SnapshotInfo, restoreSize int64) error {
+	t, err := util.ParseTimeZ(snapInfo.Created)
+	if err != nil {
+		return err
+	}
+	creationTime := metav1.NewTime(t)
+	snap.Status.Parent = snapInfo.Parent
+	snap.Status.Children = snapInfo.Children
+	snap.Status.MarkRemoved = snapInfo.Removed
+	snap.Status.UserCreated = snapInfo.UserCreated
+	snap.Status.CreationTime = &creationTime
+	snap.Status.Size = snapInfo.Size
+	snap.Status.Labels = snapInfo.Labels
+	snap.Status.RestoreSize = restoreSize
+	if snap.Status.MarkRemoved || snap.DeletionTimestamp != nil {
+		snap.Status.ReadyToUse = false
+	} else {
+		snap.Status.ReadyToUse = true
+		snap.Status.Error = ""
+	}
+	return nil
+}
+
+func (sc *SnapshotController) getTheOnlyEngineCRforSnapshot(snapshot *longhorn.Snapshot) (*longhorn.Engine, error) {
+	engines, err := sc.ds.ListVolumeEngines(snapshot.Spec.Volume)
+	if err != nil {
+		return nil, errors.Wrap(err, "getTheOnlyEngineCRforSnapshot")
+	}
+	if len(engines) != 1 {
+		return nil, fmt.Errorf("getTheOnlyEngineCRforSnapshot: found more than 1 engines for volume %v", snapshot.Spec.Volume)
+	}
```

Comment:
Is `len(engines) == 0` possible ?

Reply:
Could be the moment when the volume is being deleted. We delete the engine CR first. Anyways, it should be a temporary state